### ISSUE TITLE
macOS Safari fullscreen offset

### ISSUE THREAD

**Issue by abelbeck** [codemirror/CodeMirror#5960]
Hi,

Our project is nearing completion of adding CodeMirror support.  Works quite nicely, except...

During testing we noticed an issue that only occurs with Safari (Firefox and Chrome work perfectly) and involves using the `fullscreen` addon.  And I have a clue for a fix.

As tested: Safari 12.1.2, macOS 10.14.6

Here is the original screen:

<img width="847" alt="safari-normal" src="https://user-images.githubusercontent.com/3116543/62418392-8f91ee80-b62d-11e9-85cc-47a4ade027de.png">

Here is the screen after calling **twice** ...
```
cm.setOption("fullScreen", !cm.getOption("fullScreen"));
```

<img width="1073" alt="safari-bug" src="https://user-images.githubusercontent.com/3116543/62418395-97ea2980-b62d-11e9-8068-62233cb2da6c.png">

Clue: The above "offset" does **not** occur if the first thing I do is type `CMD-f` followed by `ESC` key (a NoOp).  Another clue, the `dialog` addon is also required to keep the bug from appearing.  If I don't load the `dialog` addon the NoOp "search" trick no longer fixes the problem.

Then I thought something in `addon/dialog/dialog.js` was poking at Safari properly, so I tried calling:
```
cm.openNotification("Lonnie");
```
but that did not fix the issue as performing `CMD-f` followed by `ESC` key did.

**Any insight on what the Safari problem might be ?**

Below are some additional snippets:

Loaded CodeMirror files:
```
<link rel="stylesheet" href="/common/codemirror/lib/codemirror.css" type="text/css" />
<link rel="stylesheet" href="/common/codemirror/addon/dialog/dialog.css" type="text/css" />
<link rel="stylesheet" href="/common/codemirror/addon/display/fullscreen.css" type="text/css" />
<script language="JavaScript" type="text/javascript" src="/common/codemirror/lib/codemirror.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/addon/search/search.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/addon/search/searchcursor.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/addon/dialog/dialog.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/addon/display/fullscreen.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/mode/asterisk/asterisk.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/mode/properties/properties.js"></script>
<script language="JavaScript" type="text/javascript" src="/common/codemirror/mode/shell/shell.js"></script>
```

Keys to control `fullscreen` mode:
```
    // Tab key toggles fullscreen, Esc returns
    cm.setOption("extraKeys", {
      Tab: function(cm) {
        cm.setOption("fullScreen", !cm.getOption("fullScreen"));
      },
      Esc: function(cm) {
        if (cm.getOption("fullScreen")) cm.setOption("fullScreen", false);
      }
    });
```

**Comment by marijnh** [comment]
I didn't manage to reproduce this. You can probably use the browser's dev tools to figure out where the extra space is coming from, and use that to narrow this down further.

**Comment by abelbeck** [comment]
In our case, the `textarea` is wrapped in a table:
```
<table class="stdtable"><tr><td>
<textarea id="ed" name="edit_text" rows="30" cols="95" wrap="off" class="editText">...
...
```
It looks like on exit from `fullscreen` the textarea/codemirror is no longer associated with the table.

Another snippet:
```
    var ta = document.getElementById("ed");
    var cm = CodeMirror.fromTextArea(ta, {
      lineNumbers: true
    });
```
Sadly, I'm not a Safari Develop menu guru.

**Comment by abelbeck** [comment]
Looking deeper, I'm convinced this issue is a Safari bug/misfeature rather than a CodeMirror issue.

It turns out for our case the `textarea` is nested within two `table` tags and Safari doesn't appear to keep track of things as it should when `fullscreen` is exited.

The fact that `CMD-f` `ESC` in CodeMirror "fixes" the issue is still interesting.

As such I'm closing this issue.

**Comment by abelbeck** [comment]
I think I have solved this issue with desktop Safari.

Basically for fullscreen to work in Safari, when focus changes to CodeMirror there must be an element in focus, else the display offset described above will occur.

In our case the first element is a `select` list of files to edit, and by default no element was in focus so clicking on the TextArea would have this issue.  By simply clicking the file list and then clicking on the TextArea solved the problem.

Step one of the solution:
```
    window.onload = function() {
      document.getElementById("list").focus();
    };
```
make the files list in focus by default.

This almost did it, but if a user clicked outside of the window the files list would switch to `blur`, and then if the TextArea was clicked, the same issue would occur.

Step two of the solution:
```
    // Desktop Safari wants a relatedTarget, else fullscreen exits incorrectly
    cm.on("focus", function(cm, change) {
      if (change.relatedTarget == null && window.safari !== undefined) {
        document.getElementById("list").focus();
        document.getElementById("list").blur();
      }
    });
```
toggling the list element focus/blur works around the issue when no `relatedTarget` was defined when the TextArea came into focus.

BTW, the `window.safari !== undefined` is a Desktop-only Safari test

I also tried this fix with iOS and a Bluetooth keyboard, whereby selecting the files list at onload was enough to fix the issue there. The "step two" fix was not needed on iOS and in fact caused an undesirable flash of the file list.

This also confirms why `CMD-f` `ESC` fixed things since it internally toggled focus.

Marked as solved (Safari workaround).
